From 92e7d3e1c6865d8cf7f819d867f03887c8bdda92 Mon Sep 17 00:00:00 2001
From: Shaun Tancheff <shaun@tancheff.com>
Date: Wed, 9 Mar 2016 15:54:52 -0600
Subject: [PATCH 25/29] Try, try, again.

---
 drivers/scsi/sd.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/scsi/sd.c b/drivers/scsi/sd.c
index 14805a8..2c9b4d3 100644
--- a/drivers/scsi/sd.c
+++ b/drivers/scsi/sd.c
@@ -1216,7 +1216,8 @@ static int sd_setup_zoned_cmnd(struct scsi_cmnd *cmd)
 			cmd->cmnd[0] = ATA_16;
 			cmd->cmnd[1] = (0x6 << 1) | 1;
 			cmd->cmnd[2] = 0x0e;
-			cmd->cmnd[3] = ZBC_REPORT_OPT;
+			cmd->cmnd[3] = bio_get_streamid(bio);
+			cmd->cmnd[4] = ATA_SUBCMD_REP_ZONES;
 			cmd->cmnd[4] = bio_get_streamid(bio);
 			cmd->cmnd[5] = ((nr_bytes / 512) >> 8) & 0xff;
 			cmd->cmnd[6] = (nr_bytes / 512) & 0xff;
-- 
2.7.0

