From d34e5e9e60d9686c0d5b60a4e560b41239eaed10 Mon Sep 17 00:00:00 2001
From: Hannes Reinecke <hare@suse.de>
Date: Tue, 28 Jul 2015 08:26:16 +0200
Subject: [PATCH 30/64] block: update chunk_sectors in blk_stack_limits()

Signed-off-by: Hannes Reinecke <hare@suse.com>
---
 block/blk-settings.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/block/blk-settings.c b/block/blk-settings.c
index dd49735..8c0aca4 100644
--- a/block/blk-settings.c
+++ b/block/blk-settings.c
@@ -630,6 +630,10 @@ int blk_stack_limits(struct queue_limits *t, struct queue_limits *b,
 			t->discard_granularity;
 	}
 
+	if (b->chunk_sectors)
+		t->chunk_sectors = min_not_zero(t->chunk_sectors,
+						b->chunk_sectors);
+
 	return ret;
 }
 EXPORT_SYMBOL(blk_stack_limits);
-- 
2.8.1

