From e127026b936a37d3dd9cd948f4fa4275766d9bb2 Mon Sep 17 00:00:00 2001
From: Shaun Tancheff <shaun.tancheff@seagate.com>
Date: Fri, 26 Aug 2016 20:51:54 -0500
Subject: [PATCH 5/7] Bump up a few revs to stop getting over written by
 upstream.

---
 debian/changelog | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/debian/changelog b/debian/changelog
index e627336..1795ae5 100644
--- a/debian/changelog
+++ b/debian/changelog
@@ -1,4 +1,4 @@
-util-linux (2.27.1-6ubuntu3.1) xenial; urgency=medium
+util-linux (2.27.1-9ubuntu9.1) xenial; urgency=medium
 
   * debian/patches/git_hybrid_cds.patch:
     - backport upstream fix to make hybrid disk work again, should resolve
-- 
2.9.3

