From 9ecb2dbdc9527bd449bcf7beea66b2d11e39dae7 Mon Sep 17 00:00:00 2001
From: Shaun Tancheff <shaun@tancheff.com>
Date: Fri, 2 Sep 2016 00:10:29 -0500
Subject: [PATCH 6/7] Make superblock 512 bytes in length.

---
 libblkid/src/superblocks/zdm.c | 24 +++++++++++++++++++++---
 1 file changed, 21 insertions(+), 3 deletions(-)

diff --git a/libblkid/src/superblocks/zdm.c b/libblkid/src/superblocks/zdm.c
index aaba06c..a01b4d6 100644
--- a/libblkid/src/superblocks/zdm.c
+++ b/libblkid/src/superblocks/zdm.c
@@ -19,11 +19,17 @@ static const char zdm_magic[] = {
 	0x82, 0x65, 0xf5, 0x7f, 0x48, 0xba, 0x6d, 0x81
 };
 
+/**
+ * A superblock stored at the 0-th block of a deivce used to
+ * re-create identify and manipulate a ZDM instance.
+ * Contains enough information to repeat the dmsetup magic create/restore
+ * an instance.
+ */
 struct zdm_super_block {
 	uint32_t crc32;
 	uint32_t reserved;
-	uint8_t magic[ARRAY_SIZE(zdm_magic)];
-	uuid_t  uuid;
+	uint8_t  magic[ARRAY_SIZE(zdm_magic)];
+	uuid_t   uuid;
 	uint32_t version;     /* 0xMMMMmmpt */
 	uint64_t sect_start;
 	uint64_t sect_size;
@@ -39,6 +45,19 @@ struct zdm_super_block {
 	uuid_t uuid_meta;     /* UUID of metadata block device */
 	uint32_t is_meta;     /* true if this is the metadata device */
 	uint32_t io_queue;    /* true if I/O queue should be enabled. */
+	uint32_t gc_prio_def;    /* 0xFF00 */
+	uint32_t gc_prio_low;    /* 0x7FFF */
+	uint32_t gc_prio_high;   /* 0x0400 */
+	uint32_t gc_prio_crit;   /* 0x0040 */
+	uint32_t gc_wm_crit;     /* 7 zones free */
+	uint32_t gc_wm_high;     /* < 5% zones free */
+	uint32_t gc_wm_low;      /* < 25% zones free */
+	uint32_t gc_status;      /* on/off/force */
+	uint32_t cache_size;      /* cache size (reserve) */
+	uint32_t cache_reada;     /* number of block to read-ahead */
+	uint32_t cache_ageout_ms; /* number of ms before droping */
+	uint32_t cache_to_pagecache;  /* if dropping through page-cache */
+	uint8_t  padding[264]; /* pad to 512 bytes */
 };
 typedef struct zdm_super_block zdm_super_block_t;
 
@@ -56,7 +75,6 @@ static uint32_t zdm_crc32(zdm_super_block_t *sblk)
 	return calc;
 }
 
-
 static int probe_zdm(blkid_probe pr, const struct blkid_idmag *mag)
 {
 	zdm_super_block_t *sblk;
-- 
2.9.3

